File: packages/local/internal/projectindex/staticindex/protocol/prompt_text.go
package protocol

const (
	// PromptTextMethod identifies the cache-bypassing compiler query.
	PromptTextMethod = "promptTextQuery"
	// PromptTextProtocolVersion is the Rust-to-Go JSON ABI version.
	PromptTextProtocolVersion uint16 = 1
)

// PromptTextPosition is a zero-based UTF-16 source position.
type PromptTextPosition struct {
	Line      uint32 `json:"line"`
	Character uint32 `json:"character"`
}

// PromptTextRange is a half-open UTF-16 source range.
type PromptTextRange struct {
	Start PromptTextPosition `json:"start"`
	End   PromptTextPosition `json:"end"`
}

// PromptTextOffsetRange is a half-open UTF-16 offset range within one island.
type PromptTextOffsetRange struct {
	Start uint32 `json:"start"`
	End   uint32 `json:"end"`
}

// PromptTextDocumentRevision identifies the exact open buffer analyzed.
type PromptTextDocumentRevision struct {
	OpenEpoch  uint64 `json:"openEpoch"`
	Version    int64  `json:"version"`
	SourceHash string `json:"sourceHash"`
}

// PromptTextStatusKind is the completeness discriminant used at request and
// template level.
type PromptTextStatusKind string

const (
	PromptTextStatusComplete    PromptTextStatusKind = "complete"
	PromptTextStatusTruncated   PromptTextStatusKind = "truncated"
	PromptTextStatusUnsupported PromptTextStatusKind = "unsupported"
)

// PromptTextAnalysisStatus reports whether a request or template is complete.
type PromptTextAnalysisStatus struct {
	Kind PromptTextStatusKind `json:"kind"`
}

// PromptTextLimits centralizes transient projection and output bounds.
type PromptTextLimits struct {
	MaxSourceBytes               uint32 `json:"maxSourceBytes"`
	MaxTemplates                 uint32 `json:"maxTemplates"`
	MaxTemplateBytes             uint32 `json:"maxTemplateBytes"`
	MaxTraversalNodes            uint32 `json:"maxTraversalNodes"`
	MaxOutputBytes               uint32 `json:"maxOutputBytes"`
	MaxStringRefactors           uint32 `json:"maxStringRefactors"`
	MaxStringRefactorBytes       uint32 `json:"maxStringRefactorBytes"`
	MaxStringRefactorOutputBytes uint32 `json:"maxStringRefactorOutputBytes"`
	MaxFragments                 uint32 `json:"maxFragments"`
	MaxFragmentJoins             uint32 `json:"maxFragmentJoins"`
	MaxFragmentBytes             uint32 `json:"maxFragmentBytes"`
	MaxFragmentDepth             uint32 `json:"maxFragmentDepth"`
	MaxPreviewBytes              uint32 `json:"maxPreviewBytes"`
}

// PromptTextFragment is one bounded, semantically proven fragment input.
type PromptTextFragment struct {
	ID         string          `json:"id"`
	Symbol     string          `json:"symbol"`
	File       string          `json:"file"`
	SourceHash string          `json:"sourceHash"`
	Range      PromptTextRange `json:"range"`
	Snippet    string          `json:"snippet"`
}

// PromptTextEvidenceProof identifies an externally verified evidence edge.
type PromptTextEvidenceProof string

const (
	// PromptTextProofSemanticExact is a current semantic fragment resolution.
	PromptTextProofSemanticExact PromptTextEvidenceProof = "semantic-exact"
)

// PromptTextInterpolationJoinKey identifies one exact interpolation occurrence.
type PromptTextInterpolationJoinKey struct {
	File            string          `json:"file"`
	SourceHash      string          `json:"sourceHash"`
	TemplateRange   PromptTextRange `json:"templateRange"`
	Interpolation   uint32          `json:"interpolation"`
	ExpressionRange PromptTextRange `json:"expressionRange"`
}

// PromptTextFragmentJoin resolves one interpolation to a supplied fragment.
type PromptTextFragmentJoin struct {
	Key        PromptTextInterpolationJoinKey `json:"key"`
	FragmentID string                         `json:"fragmentId"`
	Proof      PromptTextEvidenceProof        `json:"proof"`
}

// PromptTextQuery is one exact, cache-bypassing open-document query.
type PromptTextQuery struct {
	ProtocolVersion uint16                     `json:"protocolVersion"`
	File            string                     `json:"file"`
	LanguageID      string                     `json:"languageId"`
	Revision        PromptTextDocumentRevision `json:"revision"`
	Source          string                     `json:"source"`
	Fragments       []PromptTextFragment       `json:"fragments"`
	FragmentJoins   []PromptTextFragmentJoin   `json:"fragmentJoins"`
	Limits          PromptTextLimits           `json:"limits"`
}

// PromptTextWorkerRequest wraps one query for the persistent compiler process.
type PromptTextWorkerRequest struct {
	ID     uint64          `json:"id"`
	Method string          `json:"method"`
	Query  PromptTextQuery `json:"query"`
}

// PromptTextRefactorProofLevel identifies compiler-owned proof strength.
type PromptTextRefactorProofLevel string

const (
	PromptTextRefactorProofSyntaxExact PromptTextRefactorProofLevel = "syntax-exact"
)

// PromptTextRefactorProof is one exact ordinary-string conversion proof.
type PromptTextRefactorProof struct {
	Kind         string                       `json:"kind"`
	CandidateID  uint32                       `json:"candidateId"`
	Range        PromptTextRange              `json:"range"`
	ExpectedText string                       `json:"expectedText"`
	TemplateText string                       `json:"templateText"`
	Proof        PromptTextRefactorProofLevel `json:"proof"`
}

// PromptTextRefactorAnalysis has limits and completeness independent from
// tagged-template analysis.
type PromptTextRefactorAnalysis struct {
	Status PromptTextAnalysisStatus  `json:"status"`
	Proofs []PromptTextRefactorProof `json:"proofs"`
}

// PromptTextQueryResponse is one normalized result for an exact revision.
type PromptTextQueryResponse struct {
	ProtocolVersion uint16                     `json:"protocolVersion"`
	File            string                     `json:"file"`
	Revision        PromptTextDocumentRevision `json:"revision"`
	Status          PromptTextAnalysisStatus   `json:"status"`
	Templates       []PromptTextTemplate       `json:"templates"`
	Refactors       PromptTextRefactorAnalysis `json:"refactors"`
}

File: crates/protocol/src/prompt_text.rs
//! Data-only contracts for transient PromptText analysis.
//!
//! These records cross the persistent native-worker boundary. They contain
//! normalized UTF-16 evidence only: parser ASTs, dirty facts, and cache state
//! must never enter this module.

use serde::{Deserialize, Serialize};

mod preview;
mod structure;

pub use preview::*;
pub use structure::*;

/// Persistent-worker method for one cache-bypassing PromptText query.
pub const PROMPT_TEXT_QUERY_METHOD: &str = "promptTextQuery";

/// JSON ABI version shared by Rust, Go, LSP, and the VS Code client.
pub const PROMPT_TEXT_PROTOCOL_VERSION: u16 = 1;

/// A zero-based UTF-16 position.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextPosition {
    pub line: u32,
    pub character: u32,
}

/// A half-open range measured in zero-based UTF-16 positions.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextRange {
    pub start: PromptTextPosition,
    pub end: PromptTextPosition,
}

/// A half-open UTF-16 code-unit range within one projected literal island.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextOffsetRange {
    pub start: u32,
    pub end: u32,
}

/// Exact editor-buffer revision analyzed by the transient compiler.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextDocumentRevision {
    pub open_epoch: u64,
    pub version: i64,
    pub source_hash: String,
}

/// Completeness of a whole request or one included template payload.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "kind",
    rename_all = "kebab-case",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum PromptTextAnalysisStatus {
    Complete,
    Truncated,
    Unsupported,
}

/// Centralized bounds applied by projection, classification, and serialization.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextLimits {
    pub max_source_bytes: u32,
    pub max_templates: u32,
    pub max_template_bytes: u32,
    pub max_traversal_nodes: u32,
    pub max_output_bytes: u32,
    pub max_string_refactors: u32,
    pub max_string_refactor_bytes: u32,
    pub max_string_refactor_output_bytes: u32,
    pub max_fragments: u32,
    pub max_fragment_joins: u32,
    pub max_fragment_bytes: u32,
    pub max_fragment_depth: u32,
    pub max_preview_bytes: u32,
}

/// One bounded, semantically proven fragment available to static analysis.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextFragment {
    pub id: String,
    pub symbol: String,
    pub file: String,
    pub source_hash: String,
    pub range: PromptTextRange,
    pub snippet: String,
}

/// Strength of one externally supplied interpolation-to-fragment proof.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PromptTextEvidenceProof {
    SemanticExact,
}

/// Exact current-source occurrence receiving a proven fragment.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextInterpolationJoinKey {
    pub file: String,
    pub source_hash: String,
    pub template_range: PromptTextRange,
    pub interpolation: u32,
    pub expression_range: PromptTextRange,
}

/// One semantic-exact edge from an interpolation to a supplied fragment.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextFragmentJoin {
    pub key: PromptTextInterpolationJoinKey,
    pub fragment_id: String,
    pub proof: PromptTextEvidenceProof,
}

/// One cache-bypassing query over an exact open-document revision.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextQueryRequest {
    pub protocol_version: u16,
    pub file: String,
    pub language_id: String,
    pub revision: PromptTextDocumentRevision,
    pub source: String,
    pub fragments: Vec<PromptTextFragment>,
    pub fragment_joins: Vec<PromptTextFragmentJoin>,
    pub limits: PromptTextLimits,
}

/// Persistent-worker envelope for one PromptText query.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextWorkerRequest {
    pub id: u64,
    pub method: String,
    pub query: PromptTextQueryRequest,
}

/// Strength of one compiler-owned ordinary-string conversion proof.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PromptTextRefactorProofLevel {
    SyntaxExact,
}

/// One exact ordinary-string replacement proven independently of template
/// classification.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "kind",
    rename_all = "kebab-case",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum PromptTextRefactorProof {
    OrdinaryStringToMd {
        candidate_id: u32,
        range: PromptTextRange,
        expected_text: String,
        template_text: String,
        proof: PromptTextRefactorProofLevel,
    },
}

/// Independent completeness and source-order proofs for string refactors.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextRefactorAnalysis {
    pub status: PromptTextAnalysisStatus,
    pub proofs: Vec<PromptTextRefactorProof>,
}

/// Normalized result for one exact request revision.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PromptTextQueryResponse {
    pub protocol_version: u16,
    pub file: String,
    pub revision: PromptTextDocumentRevision,
    pub status: PromptTextAnalysisStatus,
    pub templates: Vec<PromptTextTemplate>,
    pub refactors: PromptTextRefactorAnalysis,
}
